### PR TITLE
close #6 feature to terminate job earlier

### LINKED ISSUE #6: EigenFlow should support a feature to terminate a job early
EigenFlow should have a feature that allows a job to terminate early.  This feature would allow a job to determine that it has no need to continue through the later phases (for example in a double load scenario).  In this case the job should be able to specify that rather than failing we just want to terminate the run early, and then the next run can continue as normal.

### PR DESCRIPTION
**Pull request by dmitri-carpov**

### CODE REVIEW

**Comment by suhailshergill** [comment]
@dmitri-carpov please break the commit into smaller bites, otherwise it would be difficult for me to review

File: doc/error-handling.md [hunk 1]
```diff
@@ -1,11 +1,11 @@
-# Error Handling
+### Error Handling
```
Comment (jonas) [inline doc/error-handling.md]: This should be a top-level heading

File: doc/error-handling.md [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
 
 By default, if no strategy is defined, when an exception happens the process switches to the `failed` stage remembering
-the stage it was trying to execute and exits.
+the stage it was trying to execute and exists.
```
Comment (jonas) [inline doc/error-handling.md]: This seems to revert a previous edit: the process `exits` after recording the failed stage.

File: doc/error-handling.md [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
 
-The simplest possible strategy would be: no matter what happens retry _n_-times with given interval:
+The simplest possible strategy would be: no matter what happens retry n-times with given interval:
 
-```scala
+```
```
Comment (jonas) [inline doc/error-handling.md]: reverts

File: doc/error-handling.md [hunk 4]
```diff
@@ -13,37 +13,41 @@ SomeStage {
 
 You can define multiple strategies depending on exception type:
 
-```scala
+```
```
Comment (jonas) [inline doc/error-handling.md]: We want scala

**Comment by jonas** [comment]
Most changes are from updating the license header

**Comment by suhailshergill** [comment]
@dmitri-carpov looks good (:+1: from me). that being said, since this feature is blocking a few items i think it would be worthwhile for @WadeWaldron also takes a brief look to vet things

**Comment by jonas** [comment]
Minor nit: new code doesn't use `src/main/com.mediative.eigenflow` but `src/main/com/mediative/eigenflow`

File: src/it/resources/application.conf [hunk 5]
```diff
@@ -0,0 +1,8 @@
+process {
+  id = "test"
+}
+
+akka {
+  log-dead-letters = 0
+  loglevel = "ERROR"
```
Comment (WadeWaldron) [inline src/it/resources/application.conf:7]: I wonder if this should be changed to "WARNING"?

File: src/main/scala/com.mediative.eigenflow/domain/RecoveryStrategy.scala [hunk 6]
```diff
@@ -18,19 +18,16 @@ package com.mediative.eigenflow.domain
...
 import scala.concurrent.duration.FiniteDuration
 
-/**
- * Retry strategy.
- *
- * @param interval Interval between retries.
- * @param attempts Number of retries.
- */
-case class Retry(interval: FiniteDuration, attempts: Int)
+trait RecoveryStrategy
```
Comment (WadeWaldron) [inline src/main/scala/com.mediative.eigenflow/domain/RecoveryStrategy.scala]: I suggest making this a sealed trait.

File: src/main/scala/com.mediative.eigenflow/domain/RecoveryStrategy.scala [hunk 7]
```diff
@@ -18,19 +18,16 @@ package com.mediative.eigenflow.domain
...
+
+object RecoveryStrategy {
+  case class Retry(interval: FiniteDuration, attempts: Int) extends RecoveryStrategy
 
-object Retry {
+  case object Fail extends RecoveryStrategy
 
-  import scala.concurrent.duration._
+  case object Complete extends RecoveryStrategy
```
Comment (WadeWaldron) [inline src/main/scala/com.mediative.eigenflow/domain/RecoveryStrategy.scala:28]: I wonder if "Complete" should have some kind of additional information (eg. reason: String)?

File: src/it/resources/application.conf (hunk 5 above)
Comment (dmitri-carpov) [inline src/it/resources/application.conf:7]: `eigenflow` uses warnings to report retries, what produces too much noise during testing.
we could potentially improve it by setting log level only for `eigenflow` package.

Comment (WadeWaldron) [inline src/it/resources/application.conf:7]: Ah sorry, I didn't realize this was in the "it" resources.

Also, did you mean "eigenflow uses warnings to report retries"?  That would seem more appropriate.

File: src/main/scala/com.mediative.eigenflow/domain/RecoveryStrategy.scala (hunk 7 above)
Comment (dmitri-carpov) [inline src/main/scala/com.mediative.eigenflow/domain/RecoveryStrategy.scala:28]: When it switches to `Complete` stage it passes `failure.message` as the stage message, thus the reason should be obvious from the exception, unless exception's message was not informative.

Comment (WadeWaldron) [inline src/main/scala/com.mediative.eigenflow/domain/RecoveryStrategy.scala:28]: Let's start with that.  We can always add more later if we decide it's not informative enough.

File: src/main/scala/com.mediative.eigenflow/domain/RecoveryStrategy.scala (hunk 6 above)
Comment (suhailshergill) [inline src/main/scala/com.mediative.eigenflow/domain/RecoveryStrategy.scala]: whoops, missed that. indeed.

File: src/it/resources/application.conf (hunk 5 above)
Comment (dmitri-carpov) [inline src/it/resources/application.conf:7]: yes, warnings :)

**Pull request merged**